Update database setup guide (#192)
scalar-labs/scalar-kubernetes: docs/SetupDatabaseForAWS.md

# Set up a database for ScalarDB/ScalarDL deployment on AWS

This guide explains how to set up a database for ScalarDB/ScalarDL deployment on AWS.

## Amazon DynamoDB

### Authentication method

When you use DynamoDB, you must set `REGION`, `ACCESS_KEY_ID`, and `SECRET_ACCESS_KEY` in the ScalarDB/ScalarDL properties file as follows.

```properties
scalar.db.contact_points=<REGION>
scalar.db.username=<ACCESS_KEY_ID>
scalar.db.password=<SECRET_ACCESS_KEY>
scalar.db.storage=dynamo
```

Please refer to the following document for more details on the properties for DynamoDB.

* [Getting Started with ScalarDB on DynamoDB](https://github.com/scalar-labs/scalardb/blob/master/docs/getting-started-with-scalardb-on-dynamodb.md)

### Required configuration/steps

DynamoDB is available for use in AWS by default. You do not need to set up anything manually to use it.

### Optional configurations/steps

> #### Enable pint in time recovery (Recommended in the production environment)

Reviewer: ```suggestion
#### Enable point-in-time recovery (Recommended in the production environment)
```

Author: Thank you for the suggestion!
I applied.

> You can enable PITR as a backup/restore method for DynamoDB. If you use [ScalarDB Schema Loader](https://github.com/scalar-labs/scalardb/blob/master/docs/schema-loader.md) for creating schema, it enables the PITR feature for tables by default. Please refer to the official document for more details.

Reviewer: I think it would be helpful if we could say why it should be enabled.
Also, it would be more user-friendly if we could add the following link somewhere.
https://github.com/scalar-labs/scalardb/blob/master/docs/backup-restore.md

The same comment applies to the sections of the other databases in the doc.

* [Point-in-time recovery for DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/PointInTimeRecovery.html)

#### Configure monitoring (Recommended in the production environment)

You can configure the monitoring and logging of DynamoDB using its native feature. Please refer to the official document for more details.

* [Monitoring and logging](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/monitoring.html)

#### Use VPC endpoint (Recommended in the production environment)

// Note that We have not yet tested this feature with Scalar products.  
// TODO: We need to test this feature with Scalar products.

> * [Using Amazon VPC endpoints to access DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/vpc-endpoints-dynamodb.html)

Author: We have not yet tested this feature with Scalar products.
I will do the test and update the document as another task.

### Optional configurations/steps

#### Enable automated backups (Recommended in the production environment)

You can enable automated backups. Please refer to the official document for more details.

* [Working with backups](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_WorkingWithAutomatedBackups.html)

#### Configure monitoring (Recommended in the production environment)

You can configure the monitoring and logging of RDS using its native feature. Please refer to the official documents for more details.

* [Monitoring metrics in an Amazon RDS instance](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Monitoring.html)
* [Monitoring events, logs, and streams in an Amazon RDS DB instance](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Monitor_Logs_Events.html)

#### Disable public access (Recommended in the production environment)

Public access is disabled by default. You can access the RDS database instance from the Scalar product pods on your EKS cluster as follows.

* Create the RDS database instance on the same VPC as your EKS cluster.

> * Connect the VPC for the RDS and the VPC for the EKS cluster for the Scalar product deployment using [VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html). (// TODO: We need to test this feature with Scalar products.)

Author: We have not yet tested this feature with Scalar products.
I will do the test and update the document as another task.

If you want to change some backup configurations like the backup retention period and backup window, you can configure them. Please refer to the official document for more details.

* [Backing up and restoring an Amazon Aurora DB cluster](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/BackupRestoreAurora.html)

#### Configure monitoring (Recommended in the production environment)

You can configure the monitoring and logging of Amazon Aurora using its native feature. Please refer to the official documents for more details.

* [Monitoring metrics in an Amazon Aurora cluster](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/MonitoringAurora.html)
* [Monitoring events, logs, and streams in an Amazon Aurora DB cluster](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/CHAP_Monitor_Logs_Events.html)

#### Disable public access (Recommended in the production environment)

Public access is disabled by default. You can access the Amazon Aurora DB cluster from the Scalar product pods on your EKS cluster as follows.

* Create the Amazon Aurora DB cluster on the same VPC as your EKS cluster.

> * Connect the VPC for the Amazon Aurora DB cluster and the VPC for the EKS cluster for the Scalar product deployment using [VPC peering](https://docs.aws.amazon.com/vpc/latest/peering/what-is-vpc-peering.html). (// TODO: We need to test this feature with Scalar products.)

Author: We have not yet tested this feature with Scalar products.
I will do the test and update the document as another task.

```properties
scalar.db.contact_points=<REGION>
scalar.db.username=<ACCESS_KEY_ID>
scalar.db.password=<SECRET_ACCESS_KEY>
scalar.db.storage=dynamo
```

Please refer to the following document for more details on the properties for DynamoDB.

* [Getting Started with ScalarDB on DynamoDB](https://github.com/scalar-labs/scalardb/blob/master/docs/getting-started-with-scalardb-on-dynamodb.md)

### Required configuration/steps

DynamoDB is available for use in AWS by default. You do not need to set up anything manually to use it.

### Optional configurations/steps

#### Enable point-in-time recovery (Recommended in the production environment)

You can enable PITR as a backup/restore method for DynamoDB. If you use [ScalarDB Schema Loader](https://github.com/scalar-labs/scalardb/blob/master/docs/schema-loader.md) for creating schema, it enables the PITR feature for tables by default. Please refer to the official document for more details.

* [Point-in-time recovery for DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/PointInTimeRecovery.html)

> It is recommended since the point-in-time recovery feature automatically and continuously gets backups so that you can reduce the downtime (pause duration) of backup operations. Please refer to the following document for more details on how to backup/restore Scalar product data.

Reviewer: ```suggestion
It is recommended since the point-in-time recovery feature automatically and continuously takes backups so that you can reduce downtime (pause duration) for backup operations. Please refer to the following document for more details on how to backup/restore Scalar product data.
```

### Authentication method

When you use RDS, you must set `JDBC_URL`, `USERNAME`, and `PASSWORD` in the ScalarDB/ScalarDL properties file as follows.

```properties
scalar.db.contact_points=<JDBC_URL>
scalar.db.username=<USERNAME>
scalar.db.password=<PASSWORD>
scalar.db.storage=jdbc
```

Please refer to the following document for more details on the properties for RDS (JDBC databases).

* [Getting Started with ScalarDB on JDBC databases](https://github.com/scalar-labs/scalardb/blob/master/docs/getting-started-with-scalardb-on-jdbc.md)

### Required configuration/steps

#### Create an RDS database instance

You must create an RDS database instance. Please refer to the official document for more details.

* [Configuring an Amazon RDS DB instance](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_RDS_Configuring.html)

### Optional configurations/steps

#### Enable automated backups (Recommended in the production environment)

You can enable automated backups. Please refer to the official document for more details.

* [Working with backups](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_WorkingWithAutomatedBackups.html)

> It is recommended since the automated backups feature enables a point-in-time recovery feature. It can recover data to a specific point in time. It can reduce the downtime (pause duration) of backup operations when you use multi databases under Scalar products. Please refer to the following document for more details on how to backup/restore the Scalar product data.

Reviewer: ```suggestion
It is recommended since the automated backups feature enables a point-in-time recovery feature. It can recover data to a specific point in time. It can reduce downtime (pause duration) for backup operations when you use multi databases under Scalar products. Please refer to the following document for more details on how to backup/restore the Scalar product data.
```
